### ISSUE TITLE
Clarify, rename or make clear that getWriteApi will make a new thread

### ISSUE THREAD

**Issue by ZakShearman** [influxdata/influxdb-client-java#167]
It's not very clear that calling getWriteApi unnecessarily (e.g in a repeated loop) is a bad idea when really, it is as a new thread will be created each time (a RxNewThreadScheduler). Only recently noticed this in my project when I ran a thread dump and I had over 1k threads.

Just to put some abusive cases as examples:
  - https://github.com/donghoon-khan/kubernetes-demo/blob/master/app/spring-influx2/src/main/java/com/donghoon/khan/influxdemo/repository/TemperatureRepository.java
  - (not a functioning project but still a mistake made considering it would be run as a cronjob) https://github.com/Ccccx/cloud/blob/master/influxdb/src/main/java/com/example/influxdb/task/DemoTask.java
  - From what I can tell here, a new one would be made every 5 seconds https://github.com/Iniuriarum/Iniuriarum/blob/master/src/main/java/com/remcobuddelmeijer/influxdb/local/LocalInfluxDBRepository.java

Basically, it's a mistake people are making and it's not evident until you have the issue (and have actually realised you've got the issue) and have to guess where the threads are coming from.

**Comment by bednar** [comment]
@ZakShearman thanks for issue, we will take a look

**Comment by jjathman** [comment]
Along these same lines I think an example usage with Spring would be good (or other dependency injection framework). It's not really clear from the API which instances should be created by the DI framework and injected, versus used directly via method calls. I think that the WriteApi should be created once and reused but I'm not completely sure.

**Comment by zhangxianwei2015** [comment]
Hi,

I followed the demo "[**Spring Integration for InfluxDB 2.0**](https://github.com/influxdata/influxdb-client-java/tree/master/spring)" to get influxDBClient instance.

    @Autowired
    private InfluxDBClient influxDBClient;

but when use influxDBClient to query or write, thread **RxNewThreadScheduler** keep increasing.

do we have a solution to avoid this issue?

**Comment by zhangxianwei2015** [comment]
Bad document.

It really need to tell people that **RxNewThreadScheduler** will keep increasing when create WriteApi(you may get "unable to create native thread issue").

I reuse the writeApi and it seems ok now, but have the question, will resue it cause other issue?

**Comment by bednar** [comment]
@zhangxianwei2015 the reusing `WriteAPI` is correct usecase

**Comment by strayfarmer** [comment]
I reused the WriteAPI and it seems like having an issue; 
An exception(
[RxNewThreadScheduler-3] ERROR WriteErrorEvent - The error occurred during writing of data
com.influxdb.exceptions.InfluxException: interrupted
at com.influxdb.client.internal.AbstractWriteClient.toInfluxException(AbstractWriteClient.java:546)
...
) was logged yesterday. The records needs to be write into influxdb was lost from then on and i have not got any other logged exceptions . After my restarting application,  the reusing  WriteAPI worked well again.
 I wonder if there are best practices about the reusing WriteAPI to ensure records are recorded credibly. Or what should i do to avoid this phenomenon.

**Comment by bednar** [comment]
@strayfarmer, Can you share the full stacktrace of `InfluxException`?

**Comment by strayfarmer** [comment]
@bednar Ok.
<------------------>
...
16:52:12.099 [RxNewThreadScheduler-1] WARN  WriteRetriableErrorEvent - The retriable error occurred during writing of data. Reason: 'timeout'. Retry in: 5s.
16:52:17.183 [RxNewThreadScheduler-3] ERROR WriteErrorEvent - The error occurred during writing of data
com.influxdb.exceptions.InfluxException: interrupted
        at com.influxdb.client.internal.AbstractWriteClient.toInfluxException(AbstractWriteClient.java:546)
        at com.influxdb.client.internal.AbstractWriteClient.lambda$new$12(AbstractWriteClient.java:178)
        at io.reactivex.internal.subscribers.LambdaSubscriber.onNext(LambdaSubscriber.java:65)
        at io.reactivex.internal.operators.flowable.FlowableDoFinally$DoFinallySubscriber.onNext(FlowableDoFinally.java:84)
        at io.reactivex.internal.operators.mixed.FlowableConcatMapMaybe$ConcatMapMaybeSubscriber.drain(FlowableConcatMapMaybe.java:284)
        at io.reactivex.internal.operators.mixed.FlowableConcatMapMaybe$ConcatMapMaybeSubscriber.innerSuccess(FlowableConcatMapMaybe.java:179)
        at io.reactivex.internal.operators.mixed.FlowableConcatMapMaybe$ConcatMapMaybeSubscriber$ConcatMapMaybeObserver.onSuccess(FlowableConcatMapMaybe.java:322)
        at io.reactivex.internal.operators.maybe.MaybeOnErrorNext$OnErrorNextMaybeObserver.onSuccess(MaybeOnErrorNext.java:88)
        at io.reactivex.internal.operators.maybe.MaybePeek$MaybePeekObserver.onSuccess(MaybePeek.java:122)
        at io.reactivex.internal.operators.maybe.MaybeMap$MapMaybeObserver.onSuccess(MaybeMap.java:89)
        at io.reactivex.internal.operators.flowable.FlowableSingleMaybe$SingleElementSubscriber.onComplete(FlowableSingleMaybe.java:104)
        at io.reactivex.subscribers.SerializedSubscriber.onComplete(SerializedSubscriber.java:168)
        at io.reactivex.internal.operators.flowable.FlowableRetryWhen$RetryWhenSubscriber.onComplete(FlowableRetryWhen.java:82)
        at io.reactivex.internal.subscriptions.DeferredScalarSubscription.complete(DeferredScalarSubscription.java:134)
        at io.reactivex.internal.operators.maybe.MaybeToFlowable$MaybeToFlowableSubscriber.onSuccess(MaybeToFlowable.java:70)
        at io.reactivex.internal.operators.maybe.MaybeMap$MapMaybeObserver.onSuccess(MaybeMap.java:89)
        at io.reactivex.internal.operators.maybe.MaybeMap$MapMaybeObserver.onSuccess(MaybeMap.java:89)
        at io.reactivex.internal.operators.maybe.MaybeFromCallable.subscribeActual(MaybeFromCallable.java:61)
        at io.reactivex.Maybe.subscribe(Maybe.java:4262)
        at io.reactivex.internal.operators.maybe.MaybeMap.subscribeActual(MaybeMap.java:40)
        at io.reactivex.Maybe.subscribe(Maybe.java:4262)
        at io.reactivex.internal.operators.maybe.MaybeMap.subscribeActual(MaybeMap.java:40)
        at io.reactivex.Maybe.subscribe(Maybe.java:4262)
        at io.reactivex.internal.operators.maybe.MaybeToFlowable.subscribeActual(MaybeToFlowable.java:45)
        at io.reactivex.Flowable.subscribe(Flowable.java:14805)
        at io.reactivex.Flowable.subscribe(Flowable.java:14752)
        at io.reactivex.internal.operators.flowable.FlowableRepeatWhen$WhenReceiver.onNext(FlowableRepeatWhen.java:100)
        at io.reactivex.internal.operators.flowable.FlowableFlatMap$MergeSubscriber.tryEmit(FlowableFlatMap.java:282)
        at io.reactivex.internal.operators.flowable.FlowableFlatMap$InnerSubscriber.onNext(FlowableFlatMap.java:663)
        at io.reactivex.subscribers.SerializedSubscriber.onNext(SerializedSubscriber.java:100)
        at io.reactivex.internal.operators.flowable.FlowableDelay$DelaySubscriber$OnNext.run(FlowableDelay.java:114)
        at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:66)
        at io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
		at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.io.InterruptedIOException: interrupted
        at okio.Timeout.throwIfReached(Timeout.kt:98)
        at okio.OutputStreamSink.write(JvmOkio.kt:50)
        at okio.AsyncTimeout$sink$1.write(AsyncTimeout.kt:103)
        at okio.RealBufferedSink.flush(RealBufferedSink.kt:247)
        at okhttp3.internal.http1.Http1ExchangeCodec.finishRequest(Http1ExchangeCodec.kt:155)
        at okhttp3.internal.connection.Exchange.finishRequest(Exchange.kt:91)
        at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.kt:76)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:100)
        at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.kt:34)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:100)
        at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.kt:96)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:100)
        at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.kt:83)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:100)
        at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.kt:76)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:100)
        at com.influxdb.client.internal.GzipInterceptor.intercept(GzipInterceptor.java:91)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:100)
        at com.influxdb.client.internal.AuthenticateInterceptor.intercept(AuthenticateInterceptor.java:97)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:100)
        at okhttp3.logging.HttpLoggingInterceptor.intercept(HttpLoggingInterceptor.kt:152)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:100)
        at com.influxdb.internal.UserAgentInterceptor.intercept(UserAgentInterceptor.java:60)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:100)
        at okhttp3.internal.connection.RealCall.getResponseWithInterceptorChain$okhttp(RealCall.kt:197)
        at okhttp3.internal.connection.RealCall.execute(RealCall.kt:148)
        at retrofit2.OkHttpCall.execute(OkHttpCall.java:204)
        at io.reactivex.internal.operators.maybe.MaybeMap$MapMaybeObserver.onSuccess(MaybeMap.java:82)
        at io.reactivex.internal.operators.maybe.MaybeFromCallable.subscribeActual(MaybeFromCallable.java:61)
        at io.reactivex.Maybe.subscribe(Maybe.java:4262)
        at io.reactivex.internal.operators.maybe.MaybeMap.subscribeActual(MaybeMap.java:40)
        at io.reactivex.Maybe.subscribe(Maybe.java:4262)
        at io.reactivex.internal.operators.maybe.MaybeMap.subscribeActual(MaybeMap.java:40)
        at io.reactivex.Maybe.subscribe(Maybe.java:4262)
        at io.reactivex.internal.operators.maybe.MaybeToFlowable.subscribeActual(MaybeToFlowable.java:45)
        at io.reactivex.Flowable.subscribe(Flowable.java:14805)
		at io.reactivex.Flowable.subscribe(Flowable.java:14752)
        at io.reactivex.internal.operators.flowable.FlowableRepeatWhen$WhenReceiver.onNext(FlowableRepeatWhen.java:100)
        at io.reactivex.internal.operators.flowable.FlowableRetryWhen.subscribeActual(FlowableRetryWhen.java:62)
        at io.reactivex.Flowable.subscribe(Flowable.java:14805)
        at io.reactivex.internal.operators.flowable.FlowableSingleMaybe.subscribeActual(FlowableSingleMaybe.java:34)
        at io.reactivex.Maybe.subscribe(Maybe.java:4262)
        at io.reactivex.internal.operators.maybe.MaybeMap.subscribeActual(MaybeMap.java:40)
        at io.reactivex.Maybe.subscribe(Maybe.java:4262)
        at io.reactivex.internal.operators.maybe.MaybePeek.subscribeActual(MaybePeek.java:56)
        at io.reactivex.Maybe.subscribe(Maybe.java:4262)
        at io.reactivex.internal.operators.maybe.MaybeOnErrorNext.subscribeActual(MaybeOnErrorNext.java:46)
        at io.reactivex.Maybe.subscribe(Maybe.java:4262)
        at io.reactivex.internal.operators.mixed.FlowableConcatMapMaybe$ConcatMapMaybeSubscriber.drain(FlowableConcatMapMaybe.java:276)
        ... 34 common frames omitted
...
16:56:01.099 [RxNewThreadScheduler-1] WARN  WriteRetriableErrorEvent - The retriable error occurred during writing of data. Reason: 'timeout'. Retry in: 5s.
16:56:16.138 [RxNewThreadScheduler-4] WARN  WriteRetriableErrorEvent - The retriable error occurred during writing of data. Reason: 'timeout'. Retry in: 25s.
16:56:51.170 [RxNewThreadScheduler-5] WARN  WriteRetriableErrorEvent - The retriable error occurred during writing of data. Reason: 'timeout'. Retry in: 125s.
...
<------------------>
And there is my dependency version in maven: 
influxdb-client-java -> 1.15.0
influxdb-spring -> 1.15.0

**Comment by fenjen** [comment]
I have the [exact same problem](https://github.com/influxdata/influxdb-client-java/issues/168#issuecomment-843206637) (Exception message "interrupted" and no more data sent to InfluxDB server). I came accross this thread because I wanted to workaround the problem by calling `getWriteApi()` every 5 minutes or so. But I don't know if that would create infinitely threads in the long run.

**Comment by bednar** [comment]
@fenjen, please use `WriteApi` as a singleton and dispose API by `close` at the end of your program.

**Comment by fenjen** [comment]
@bednar Thanks. Then I would have to restart my application once the Influx service is not available for a short time. Unfortunately I can't do that because my process needs a long time to start up.

**Comment by bednar** [comment]
@fenjen you can configure retry strategy by properties (see docs: https://github.com/influxdata/influxdb-client-java/tree/master/client#writes) to avoid short Influx server outage.

**Comment by fenjen** [comment]
@bednar Hm, I'm using the defaults (maxRetryDelay=125s, maxRetryTime=180s). The time between the first logged retriable error (which I can see in Wireshark) and the last try (which I can not see in Wireshark) is about 160s in sum:
```
[...]          successful communication between Java client and InfluxDB service (visible in Wireshark)
12:34:20.000   manually stopped influxd
12:34:21.113   retriable_error; retry in 5s       - unsuccessful communication, visible in Wireshark
12:34:28.162   retriable_error; retry in 25s      - unsuccessful communication, visible in Wireshark
12:34:55.229   retriable_error; retry in 125s     - unsuccessful communication, visible in Wireshark
12:37:00.229   expected, but missing log message  - no TCP communication visible in WireShark
```
The unwritten points at 12:37:00 are around 1500. The default bufferLimit is 10,000. Still I tried these write options without any effect:
```
WriteOptions writeOptions = WriteOptions.builder()
		.maxRetryDelay((int) Duration.ofHours(1).toMillis())
		.bufferLimit(Integer.MAX_VALUE)
		.maxRetries(1000000000) // client complains about Integer.MAX_VALUE
		.build();
```
The behavior is unchanged with these write options.

**Comment by jjathman** [comment]
As another point, this is causing a thread leak for people who use the InfluxDB plugin for Jenkins: https://issues.jenkins.io/browse/JENKINS-65622. 

Even Influx's own documentation you get when using Chronograf does this incorrectly. This code will leak threads:
<img width="578" alt="Screen Shot 2021-08-04 at 12 53 58 PM" src="https://user-images.githubusercontent.com/6406/128230460-08c1fdcd-06e5-4590-8613-d92d26a602db.png">

Having a `get` method cause side effects like this is definitely not what most Java programmers would expect. The API should be `createWriteApi` to make it more clear that it has side-effects.

**Comment by bednar** [comment]
@jjathman, thanks for your information. We are preparing a new major version of the client due the #251 and I will do my best to also fix this naming in the new major version.

Thanks again for your information.

**Comment by bednar** [comment]
I have prepared a PR with clarification how to use `WriteApi` - #255. Development version of the client will be available after #255 acceptance, stable version in regular release cycle.
